### PR TITLE
Update conda-forge-pinnings, conda-build and bioconda-specific compilers

### PR DESCRIPTION
**Pull request by dpryan79**
This adds python 3.9 as well and basically updates everything else. We could try to time this to correspond to the htslib 1.12 release instead, but given the relative urgency of the TBB stuff perhaps waiting is a bad option.

### CODE REVIEW

**Comment by dpryan79** [comment]
I don't know why pandas isn't working, it's fine locally.

**Comment by dpryan79** [comment]
@mbargull Am I correct in assuming that the Linux failure is due to the hashing change?

**Comment by dpryan79** [comment]
Indeed, these are the fields that are added to the hash:

```
INFO :: The inputs making up the hashes for the built packages are as follows:
{
  "one-0.1-py36ha770c72_0": {
    "recipe": {
      "channel_targets": "conda-forge main",
      "target_platform": "linux-64"
    }
  }
}
```

I'll change the channel_targets to `bioconda main`

File: bioconda_utils/bioconda_utils-conda_build_config.yaml [hunk 1]
```diff
@@ -48,3 +48,8 @@ numpy:
   - 1.16
   - 1.16
 
+channel_sources:
+  - conda-forge,bioconda,defaults
+
+channel_targets:
```
Comment (mbargull) [inline bioconda_utils/bioconda_utils-conda_build_config.yaml:54]: Ouch... Yep, checked conda-build's source... This is now hard-coded to be included... I can't say I'm a big fan of including hashes for something which should be purely output-related :/.
(^-- Just a rant about conda-build's, not your changes.)
(My best guess they included it for their convenience since some recipes on conda-forge/AnacondaRecipes use `channel_targets` in some conditional Jinja control code so they don't have to manually sync the recipes.)

File: bioconda_utils/bioconda_utils-conda_build_config.yaml [hunk 2]
```diff
@@ -29,28 +29,27 @@ bamtools:
...
+  - cpython
+  - cpython
+  - cpython
+  - cpython
+numpy:
+  - 1.19
+  - 1.16
+  - 1.16
+  - 1.16
+  - 1.16
```
Comment (mbargull) [inline bioconda_utils/bioconda_utils-conda_build_config.yaml:49]: I wonder if we have to do a similar dance as with `python`/`r_base` above regarding https://github.com/conda/conda-build/issues/3974 ..

**Comment by mbargull** [comment]
> We could try to time this to correspond to the htslib 1.12 release instead,

Do we have any insights on for when that release is planned?

> but given the relative urgency of the TBB stuff perhaps waiting is a bad option.

`tbb`/`tbb-devel` only affect few recipes *directly*\*, right? We could also just do a minor pinning change with just the TBB constraint beforehand and do the mass-rebuild afterwards.
(\* of course, indirectly for workflows/other packages, I guess it has much more impact, seeing `bowtie2` et al. being affected.)

File: bioconda_utils/bioconda_utils-conda_build_config.yaml (hunk 2 above)
Comment (mbargull) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml:49]: Hm.. I find things around merged config + zip_keys too confusing (code is not without bugs and I don't understand all intentions).
How about we avoid all of that and do things manually? I.e., not use multiple "exclusive config files" for now, but just copy conda-forge's, comment out what we want to override and pre- or append our pinnings in the same file. Not ideal, but probably much safer...

**Comment by dpryan79** [comment]
@mbargull Should we pin libdeflate here? That's caused some issues on OSX in particular.

**Comment by mbargull** [comment]
> @mbargull Should we pin libdeflate here? That's caused some issues on OSX in particular.

I guess that makes sense, yes, adding it here might help update-pinnings-run to pick up recipes that depend on it.

File: bioconda_utils/bioconda_utils-conda_build_config.yaml (hunk 2 above)
Comment (mbargull) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml:51]: What do you think about this one? I'm rather concerned that `conda-build` will manage to collapse these list again :/.
IDK if it will work because we only duplicate an entry but not add a different one here (adding a different one made `conda-build` barf before).

Comment (mbargull) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml:51]: Oh, and some hours ago conda-forge updated to `numpy=1.17` -- so, probably best to update the conda-forge-pinnings version again ;).

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml:51]: Good to know, we might as well update then. The main issue with copying over their pinnings and modifying them is that it's easier to lose track of what our differences are. If you're really concerned maybe we merge and tag and then manually run a couple packages for testing. Since it's numpy and python we can use pybigwig as a test (it uses both with C so it needs the pinning).

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: I'll install this locally in a new environment and try this there.

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: It's good I started looking into this, we can't go above numpy 1.16 with python 2.7

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: Hmm, this doesn't want to build against python 3.9

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: This is definitely a problem. If we keep only python 2.7 and 3.9 then the zip_keys work properly, 2.7 builds use numpy 1.16 and 3.9 builds get numpy 1.19. However, once we add in python 3.6, 3.7, or 3.8, then only those python builds get used and the others ignored. They use the right numpy versions, but that's not terribly helpful.

Comment (mbargull) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: Ah, bummer. Yep, that sounds like the exact same "`conda-build` takes the intersection of values from `conda_build_config.yaml` and `bioconda-conda_build_config.yaml`"-issue.
So, for `numpy` itself we could go with the "append `.*`" workaround to get distinct sets.
As for `python_impl`: That one is used as a verbatim string (e.g., with `blabla  # [python_impl == "cpython"]` in conda-forge recipes). We don't use `python_impl` in our recipes currently, so it should generally not be important yet. And, more significantly, we only have a single `cpython` value in the configs for now, so the intersection remains just `cpython`.

TL;DR We could use `numpy: [ 1.16.*, 1.17.*, ... ]` for now. I'll add that change.

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: Good call, I'll run that locally to see if it resolves the issue.

Comment (dpryan79) [reply bioconda_utils/bioconda_utils-conda_build_config.yaml]: Yup, looks like that solved it locally

**Comment by dpryan79** [comment]
@mbargull This is ready to go from my side. Is there anything else from your end?

**Comment by jmarshall** [comment]
Can we talk some more about libdeflate?

IMHO the real problem with libdeflate is that most things that depend on it shouldn't. As of a week ago:

```
Has host and run dependency on libdeflate (and also htslib):

recipes/arriba/meta.yaml
recipes/bcftools-gtc2vcf-plugin/1.9/meta.yaml [bundles htslib somehow]
recipes/cancerit-allelecount/meta.yaml [libdeflate on host only]
recipes/demuxlet/meta.yaml [also bundles htslib somehow]
recipes/hap.py/meta.yaml [libdeflate on host only]
recipes/ivar/meta.yaml
recipes/megadepth/meta.yaml [libdeflate on host only]
recipes/pheniqs/meta.yaml
recipes/pysam/0.15.2/meta.yaml [run htslib implicit due to samtools/bcftools]
~~recipes/wiggletools/meta.yaml~~ [no longer depends on libdeflate since PR bioconda-recipes#27401]

Has host and run dependency on libdeflate (but not htslib):

recipes/crispritz/meta.yaml [run only]
recipes/htslib/meta.yaml
recipes/nanopolish/meta.yaml [run only]
recipes/pysam/meta.yaml
recipes/staden_io_lib/meta.yaml
recipes/vt/meta.yaml
```

Most likely all of those in the first category do not themselves use libdeflate, but only get it via htslib. So culling their lists of dependencies will insulate them from libdeflate changes.

In the second category, htslib and staden_io_lib truly do use libdeflate. Pysam currently does but in the near future I plan to raise a PR to revert it to using the shared library from bioconda's htslib package instead — so it would move to the first category. I am unsure of the exact status of crispritz, nanopolish, and vt.

So when bumping those in the first category to rebuild them against a pinned htslib 1.12, I would propose to check whether they have a non-htslib-based libdeflate dependence and simultaneously remove their libdeflate dependency if not.

I expect this would largely take care of the libdeflate problem!

**Comment by dpryan79** [comment]
@mbargull Yeah, we might be able to do without a libdeflate pinning, most of them predate the migration to conda-forge and fix of the run-time restrictions that it exports (or maybe it was an issue of the shared-library name change on OSX from .so to .dylib).

I think vt builds htslib internally, the others are using pysam (internal htslib build) or htslib, so we can remove them there (the libdeflate addition is to work around an old OSX issue I think).

**Comment by dpryan79** [comment]
Gah, that should have been @jmarshall! Sorry

**Comment by dpryan79** [comment]
So TLDR, I like what @jmarshall proposes.

**Comment by dpryan79** [comment]
I'll tag this tonight and run the pinning update on bulk today or tomorrow. I'll temporarily blacklist the packages mentioned by @jmarshall (except htslib, staden_io, and pysam) and handle them separately after the migration.
